Paper: Functional Uncertainty Constrained by Law and Experiment

fraserphysics opened the pull request:
consists of 1x rst file plus 5x pdf files

rcpaffenroth commented:
I enjoyed reading the paper and I personally found the analysis by way of Fisher information matrices to be particularly interesting.   This paper would make a find submission to a Physics or Applied Mathematics Journal.

However, I do have to admit that the role that Python plays, and how this  paper fits in the SciPy conference, is harder to see.   In particular, Python plays the staring role in the third paragraph in the introduction, but is not otherwise mentioned at all except for a tangential references to scipy.optimize.brentq and scipy.integrate.odeint on page 4.

Accordingly, I am left thirsting for more.  Why are the libraries mentioned in the third paragraph in the introduction particularly useful for this project?   Why was Python chosen, versus some other language?  Perhaps most importantly, are there lessons learned in this project that would help others considering Python for similar applications? 

In a similar vein, some of the mathematical presentation may be at too low of a level for many SciPy participants.  SciPy participants come from many diverse fields.  I have to admit that I was lost in the discussion on "Chapman Jouguet states" and 'Hugoniot curves" and I worry that others may have similar issues in several places in the text.

steve855 commented:
Thank you for taking the time to review our paper, we are glad to hear you enjoyed it. We will update the paper to focus more on the software development aspects of the project and expand the discussion of the two simulations so they are more appropriate for a broader audience. We have recently overhauled the structure of the F_UNLCE project so it is more modular, the changes can be seen in [this pull request](https://github.com/fraserphysics/F_UNCLE/pull/2).

We intend to apply this framework to real measurements and publish the results in a physics journal. We also hope to resolve the caveats discussed in the conclusions and publish in a uncertainty quantification or statistics journal. Another motivation in the development of this software was to learn and employ good software development practice. We will expand our discussion of this aspect of the project to the paper so that the paper is more relevant to this conference.

sarostru commented:
When are you making these changes?  I will be reviewing the paper this week, should I wait for the changes?

steve855 commented:
We will push the new changes tomorrow. I hope that is soon enough for your review schedule.

sarostru commented:
Great, I'll wait til then, thanks.

steve855 commented:
Thank you for your patience, we have updated the paper.

certik commented:
Disclaimer: I work at the same Lab as both of the authors.

Authors present a package `F_UNCLE` for inverting a physical model to estimate an unknown function for real experimental data.

I read the paper several times, but to be honest I don't understand how it works. For example, the case `Equation of State` (Fig. 1), how is the method better than cubic spline fitting? I would think that cubic spline fitting should produce much better agreement than in Fig. 1.  I thought that EOS maps temperature, density and pressure. In your data in Fig. 1, do you assume a particular temperature? You wrote that you are on an isentrope --- so entropy is constant. I thought that curve is different to isotherm (constant temperature, that I assume you have).

Overall, would it be possible to briefly describe (probably in the introduction) for general audience, how is this method better than other fitting methods like splines? And why would one want to use `F_UNCLE` rather than, say, cubic splines.

The figures could be improved.

* What are the three curves at Fig. 6 (bottom)?
* What is the line on Fig. 7 (top)? I thought that eigenvalue should be just a point, so I would expect to see 3 points on that plot. I can see a line, which I don't know what it means, and a point, I assume the eigenvalue 1, the value of it about 10e-9.

Regarding the software package, I reviewed https://github.com/fraserphysics/F_UNCLE, the latest master (00a566eb836687b6c7acd05b8b3102ca77301991). The paper says:

> Each class has a test suite to allow unit testing. Tests are collected and  run  using  [nose].

I tried that and it didn't work, so I reported the bug at https://github.com/fraserphysics/F_UNCLE/issues/5

The README says:

>     $ python scipy2016.py

But there is not `scipy2016.py` (I reported as https://github.com/fraserphysics/F_UNCLE/issues/6).

However, I was able to run `run_funcle.py`, here is the output that I got: https://gist.github.com/certik/bf0609ddc83c3aa4e3012bcaf83a33cf

sarostru commented:
This paper contains some interesting studies on 1D detonation models and an iterative Bayesian technique to model uncertainty of a particular equation of state (EOS).  They use two simpler surrogate models which may be solved to high accuracy to validate their statistical approach.  My main worry is that the large amount of domain specific material could be challenging for many readers, as scipy brings in a diverse multidisciplinary audience.  There also isn't a lot of discussion of the software or how python was important to the success of this work; except for a short note in the introduction and some mentions of packages used.

I find that I don't have a lot of clarity as to the precise aim of this paper, I would recommend the authors reword the abstract and introduction to be more focused.  
 - What are the contents of this paper?
 - What will the reader learn from reading it?
For instance in the introduction paragraphs 1 and 2 suggest that the aim is to develop an improved inverse problem solver, but then that quickly transitions into problems with Job schedulers.  It's difficult to understand what the relation between those two things are, and in what way F_UNCLE is resolving the issues.

With regards to the MAP algorithm in section 2, I couldn't see a clear explanation of how the algorithm in section 2 accomplishes the exclusion of physically impossible functions mentioned in the introduction (para 1).  Additionally, standard approaches aren't brought up so it is difficult to assess this algorithm in absence of any point of reference.  Is it proposed as an improvement?  How long does it take to converge?  Is it stable?  You mention that "the behavior of the experiments depends sensitively on the EOS function", but is your approach robust against this sensitivity? 

I think the Fisher information is supposed to show the stability, but how would it change with a different EOS, or different parameter values for the current EOS?  For the purposes of discussing the MAP algorithm, I find the detailed detonation study experiments actually detract from clarity since much of the working of the algorithm could be explored using simpler functions, e.g. f(x) = 1/x^3 instead of introducing the additional domain specific EOS function.  Additionally, it would make it easier to vary the function, so readers could have more confidence that this approach has more general utility than for this particular physical model.

In section 3, the cubic spline approximation is introduced, which seems to be a parameterized approach, so some explanation of why this parameterization does not suffer from the problem the authors mention with regards to other GPM models in the introduction seems necessary.

With regards to the use of surrogate problems, to again help readers from other disciplines see if they could use some of your techniques, it would be helpful to understand why these particular problems were chosen and how one can find simpler problems like this, or if it is a particularity of this domain.  

I'm not sure about including the PBX plot at the end, there is no information about what the PBX model looks like or the calculation that was done to get these numbers so I would recommend leaving it out, unless you want to explain it more fully in the text and relate it back to the surrogate experiments.

sarostru commented on papers/andrew_fraser/andrew_fraser.rst:
```diff
@@ -0,0 +1,788 @@
+   where in :math:`P_{i,k}` and :math:`q_{i,k}`, :math:`i` is the
+   iteration number and :math:`k` is the experiment number.
+
+#. Calculate :math:`G_i` and :math:`h_i` to express the appropriate
```
I couldn't find G and h defined previously

steve855 replied:
I updated the text to show what these symbols mean

fraserphysics commented:
Here we reply to Scott Rostrup's comments and and ask two questions
about direction for the paper.  Rostrup writes "I find that I don't
have a lot of clarity as to the precise aim of this paper".


-------------Origin and Aim of the Paper------------------------------
We derived our scipy2016 manuscript from a working document that we
use to explain (to ourselves and our colleagues) the math, physics and
code for the F_UNCLE project.  We hope that the manuscript serves to
introduce the F_UNCLE code and project, and that it will be useful in
the following ways:

1. Demonstrate/suggest using the scipy ecosystem and a surrogate for
   addressing difficult problems.  The paper might give confidence and
   general direction to scientists and engineers who must develop
   original solutions to their own problems.

2. Visually illustrate the utility of Fisher information for
   quantifying how tightly experimental data constrain models

3. Introduce the F_UNCLE code to folks whose problems are similar
   enough to ours that they can use it

4. Suggest the utility of better software practices to LANL colleagues
   who view software as a necessary and ugly evil


------------Questions About Directions for the Paper------------------
Both the code and the project are works in progress, and the paper may
be too much like a lab notebook for a journal publication.  In the
section "Conclusion, Caveats and Future Work", we even write, "we
emphasize that it is a work in progress."

We ask:

1. Are works in progress appropriate for the Scipy2016 Proceedings?

2. If so, do you recommend modifying the paper (primarily the
   introduction) to include the tone (less formality and more
   explanation of context) and some of the text of this note?


------------Aim and Status of the F_UNCLE Project--------------------
We need to quantitatively characterize the accuracy of models of
physical material properties which are based on existing theory and
experiments.  If the accuracy is inadequate, we must evaluate whether
or not proposed experiments or theoretical work will provide the
necessary information.  We have chosen to first work on the EOS of the
gas produced by detonating PBX-9501 because it is relatively simple.
In particular there is a model that defines its properties in terms of
a one dimensional function and simple constraints.

Characteristics of the work include:

1. The uncertain object is a *function*.  In principal it has an
   infinite number of degrees of freedom.  We don't know how to define
   a probability measure on sets in function space, and we don't know
   how to compare the utility of different families of parametric
   approximations.

2. Understanding the constraints on the unknown function and the
   connection between it and experimental measurements requires
   studying some detailed physics.

3. Simulations of some of the experiments run for about an hour on
   high performance computers.  The job control is unwieldy as are the
   mechanisms for expressing trial instances of the unknown functions
   and connecting them to the simulations.

So far we have worked in two parallel modes as we develop ideas and
algorithms.  We write code for a surrogate problem that runs in a
fraction of a minute on a PC, and we write code for fitting a model to
PBX-9501 in a high performance computing environment.  Our focus
shifts back and forth as we find and resolve problems.  As we have
progressed, we have found improving our software practices makes it
easier to express ideas, test them on PCs and implement them on the
HPCs.

Results so far indicate optimization can find good approximations to
the unknown functions and that analysis of Fisher information can
quantify how various experiments constrain uncertainty about different
aspects of them.  The results are encouraging but preliminary.

We plan to resolve some questions about choices of coordinates in
function space and add at least one more experiment to the analysis of
the PBX-9501 EOS.  We will also modify the F_UNCLE code so that we can
use it to analyze other physical models.  Then we plan to publish a
new estimate of the PBX-9501 EOS in the Journal of Applied Physics
(where previous work on that EOS appears), and we hope to publish our
analysis techniques in the SIAM/ASA Journal on Uncertainty
Quantification.


--------------Scott Rostrup's Points with our Comments----------------

1. Domain specific material could be challenging for many readers

   While we agree, we propose no major changes to the "Rate Stick" and
   "Gun" sections.  The graphical construction for CJ conditions that
   we describe in the "Rate Stick" section follows from the
   Rankine–Hugoniot conditions for shocks.  It would be a mistake to
   go into that much detail in the paper.  The dynamics of the gun are
   essentially F=ma.  We have a few equations on our poster that we
   believe make that clearer and we will include them in the next
   draft of the paper.

2. Not a lot of discussion of the software or how python is important to
   success

   We mention the major packages that we use and give a link to our
   software.  Going forward, we will work towards making the software
   self explanatory.

3. Reword the abstract and introduction:
   * What are the contents of this paper
   * What will the reader learn from reading it

   OK

4. Are we developing an inverse problem solver or are we solving
   problems with Job schedulers?

   Both.  We should probably mention that there is lots of work on
   inverse problems whose targets are functions, eg, images.  Our
   problem differs in the diversity of measurements and the nature of
   the constraints.

5. No clear explanation of how the algorithm enforces constraints.

   That is true.  While explaining the actual constraints and their
   implementation would require more text and detail than is
   appropriate, we will comment on what we are skipping and why.

6. No reference to standard approaches

   We are unaware of standard approaches.  We will add text that
   compares our problem to image estimation.

7. We don't address convergence or stability

   We have not studied them.  We will add comments about convergence
   and stability to our caveats section.

8. Fisher information to show stability?

   No.  Fisher information tells you how informative an experiment is.

9. Detailed detonation study experiments actually detract from clarity

   It's the problem that we are really working on.  Estimating the
   mean of a normally distributed scalar would be clear but
   uninteresting.  We believe that our surrogate experiments are as
   simple as they can be and still be instructive about the
   challenges of the real models we want to estimate.

10. Why are cubic splines not just another choice of parameters?

    In short, we don't know.

    It's question we are excited about.  In future work we will
    examine cost-benefit trade offs of various choices for coordinates
    in function space.  Our intuition is that cubic splines are good
    because they are mostly local and with a large number of knots
    they are very flexible.  We believe that the flexibility is
    balanced by the convexity constraint.  We imagine that we can
    study something like the relative efficiency of various coordinate
    systems for function space in terms of statistical bias and
    variance.

11. Why don't cubic splines suffer the problems of GPM models?

    GPM models allow non-convex solutions.  See Shirin Golchi's work
    on monotone functions and GPM.  F_UNCLE can enforce constraints
    that prohibit non-convex functions.

12. Why did we choose these particular problems to illustrate
    techniques?

    We developed these techniques to solve these problems

13. Recommend leaving out the PBX plot

    The PBX-9501 plot testifies to our interest in and progress on
    real problems.
>>>>> "SR" == Scott Rostrup <notifications@github.com> writes:

    SR> This paper contains some interesting studies on 1D detonation
    SR> models and an iterative Bayesian technique to model uncertainty
    SR> of a particular equation of state (EOS). They use two simpler
    SR> surrogate models which may be solved to high accuracy to
    SR> validate their statistical approach. My main worry is that the
    SR> large amount of domain specific material could be challenging
    SR> for many readers, as scipy brings in a diverse multidisciplinary
    SR> audience. There also isn't a lot of discussion of the software
    SR> or how python was important to the success of this work; except
    SR> for a short note in the introduction and some mentions of
    SR> packages used.

    SR> I find that I don't have a lot of clarity as to the precise aim
    SR> of this paper, I would recommend the authors reword the abstract
    SR> and introduction to be more focused.

    SR> * What are the contents of this paper?  * What will the reader
    SR> learn from reading it? For instance in the introduction
    SR> paragraphs 1 and 2 suggest that the aim is to develop an
    SR> improved inverse problem solver, but then that quickly
    SR> transitions into problems with Job schedulers. It's difficult to
    SR> understand what the relation between those two things are, and
    SR> in what way F_UNCLE is resolving the issues.

    SR> With regards to the MAP algorithm in section 2, I couldn't see a
    SR> clear explanation of how the algorithm in section 2 accomplishes
    SR> the exclusion of physically impossible functions mentioned in
    SR> the introduction (para 1). Additionally, standard approaches
    SR> aren't brought up so it is difficult to assess this algorithm in
    SR> absence of any point of reference. Is it proposed as an
    SR> improvement? How long does it take to converge? Is it stable?
    SR> You mention that "the behavior of the experiments depends
    SR> sensitively on the EOS function", but is your approach robust
    SR> against this sensitivity?

    SR> I think the Fisher information is supposed to show the
    SR> stability, but how would it change with a different EOS, or
    SR> different parameter values for the current EOS? For the purposes
    SR> of discussing the MAP algorithm, I find the detailed detonation
    SR> study experiments actually detract from clarity since much of
    SR> the working of the algorithm could be explored using simpler
    SR> functions, e.g. f(x) = 1/x^3 instead of introducing the
    SR> additional domain specific EOS function. Additionally, it would
    SR> make it easier to vary the function, so readers could have more
    SR> confidence that this approach has more general utility than for
    SR> this particular physical model.

    SR> In section 3, the cubic spline approximation is introduced,
    SR> which seems to be a parameterized approach, so some explanation
    SR> of why this parameterization does not suffer from the problem
    SR> the authors mention with regards to other GPM models in the
    SR> introduction seems necessary.

    SR> With regards to the use of surrogate problems, to again help
    SR> readers from other disciplines see if they could use some of
    SR> your techniques, it would be helpful to understand why these
    SR> particular problems were chosen and how one can find simpler
    SR> problems like this, or if it is a particularity of this domain.

    SR> I'm not sure about including the PBX plot at the end, there is
    SR> no information about what the PBX model looks like or the
    SR> calculation that was done to get these numbers so I would
    SR> recommend leaving it out, unless you want to explain it more
    SR> fully in the text and relate it back to the surrogate
    SR> experiments.

    SR> — You are receiving this because you authored the thread.  Reply
    SR> to this email directly, view it on GitHub, or mute the thread.



-- 
Andy Fraser                                            FraserPhysics
197 Taos St                                            andy@fraserphysics.com
Los Alamos, NM 87544                                   505 695-2788

sarostru commented:
Thanks for the detailed follow up, I like the direction of your comments, to your 2 questions:

> 1. Are works in progress appropriate for the Scipy2016 Proceedings? 

I'm not sure in general, but from your paper I think you have enough material already to be interesting, I was recommending to not try and emphasize the work in progress aspect as much.

> 2. If so, do you recommend modifying the paper (primarily the introduction) to include the tone (less formality and more explanation of context) and some of the text of this note?

The tone should still be formal, but focusing on explanation of context would be preferred.  One of the main considerations noted for reviewers is to keep scipy proceedings accessible to a general interdisciplinary audience, from https://github.com/scipy-conference/scipy_proceedings/blob/master/review_criteria.md : 

> Due to the interdisciplinary nature of SciPy, highly domain-specific jargon should be avoided or explained where possible.

kbarbary commented on papers/andrew_fraser/andrew_fraser.rst:
```diff
@@ -0,0 +1,788 @@
+.. math::
+   :label: eq-bayes
+
+   p(\theta|x) = \frac{p_l(x|\theta) p_p(\theta)}{\int p_l(x|\phi) p_p(x) d\phi}.
```
Should be `p_p(\phi)` in the denominator rather than `p_p(x)`

kbarbary commented on papers/andrew_fraser/andrew_fraser.rst:
```diff
@@ -0,0 +1,788 @@
+.. math::
+   :label: eq-map
+
+   \hat \theta \equiv {\operatorname*{argmax}}_{\phi} p(\theta|x).
```
Is this meant to be `argmax_\theta` rather than `argmax_\phi`? `p(\theta|x)` is not a function of `\phi` after all.

kbarbary commented:
Overall, I had much the same reaction as @sarostru that the precise aim of the paper could be made more clear. Does `F_UNCLE` implement a new optimization technique for a certain problem domain? Or does it implement a flexible parameterization of 1-d functions subject to some constraints (e.g., cubic splines) but with standard optimization techniques? Paragraph 5 of the introduction touches on the answer to this question, but it would be better to move this up to be the second paragraph of the introduction, directly after the paragraph explaining the problem you're trying to solve. For example: "In this paper, we introduce software F_UNCLE that addresses these problems by ..."

I would remove reference to the problems of running on HPC systems from the introduction (e.g., current paragraph 2), as these are not actually addressed by the paper.

The section "Fischer Information and a Sequence of Quadratic Programs" doesn't need such detailed equations. It's probably sufficient for this audience to state Bayes' theorem as
```
p(\theta|x) \propto p_l(x|\theta) p_p(\theta)
```
and then basically just state that a Taylor expansion of this posterior probability around the peak gives you the Fisher information (but with the posterior rather than the likelihood). I found myself slogging through the equations just to figure this out.

In the subsection "Iterative Optimization", it isn't clear if the optimization procedure described is a standard constrained optimization algorithm or is novel in some way. It might help the reader to put names to a few things: For example, refer to `q_i^T` as the gradient on the cost function and `P_i^T` as the Hessian.

For the surrogate problem sections, I'll just echo previous comments by saying that the less domain specific this can be made, the better. If the essentials of the problem can actually be reduced to "there's an unknown 1-d function we want to infer, and we know it is concave and monotonic", that will really sharpen the reader's focus.

Finally, as a Python user and developer interested in optimization techniques, but in a different domain, I would have liked to hear more exposition about the use of scipy functions in your optimization. For example, why was `scipy.optimize.brentq` used for the rate stick problem, and how does this optimization relate to the previously outlined optimization procedure?

steve855 commented:
Thank you to all four reviewers for taking the time to read our
paper. Here is a summary of the comments we received and our response
to each of them

## @rcpaffenroth

* It is hard to see the role python plays in this paper
* Why are the libraries used
* why was python chosen
* lessons learned for other projects
* mathematical presentation is too low level for general audience

  These comments are similar to those from @sarostru.  We addressed
  those with a detailed reply and we also edited the manuscript in
  response.

## @certik

* How is the method better than cubic spline fitting
  - There is a simulation between the model and the data.  We don't
    have x,y data that we can fit a spline to.
* Why constant temperature
  - We have added text that explains and justifies modeling isentropic
    expansion.  The key is that there is not enough time for heat
    conduction.
* Legend in figure 6
  - Done
* Testing
  - Currently broken, will fix before conference

## @sarostru

* What are the contents of the paper
  - Abstract was reworded
* What will the reader learn from reading it
  - Abstract was reworded
* How does the MAP function exclude physically impossible functions
  - Updated description of the model
* How does this compare to standard algorithms
  - See @fraserphysics response
* How does fisher information impact stability
  - See @fraserphysics response
* Why are domain specific models needed
  - See @fraserphysics response
* Why is a complex EOS function needed
  - See @fraserphysics response 
* Why are parameterized cubic spline better than parameter based approaches from introduction
  - See @fraserphysics response
* Should PBX be used
  - See @fraserphysics response
* Where are G and h defined
 - G and h express appropriate constraints.  We do not provide more
   detail in this paper, but we refer to the F_UNCLE code where we
   will provide the details.
    
## @kbarbary

* should p_p(\phi) be in the denominator
  - yes, corrected in paper
* should it be argmax(theta)
  - yes, corrected in paper
* move para 5 to be para 2
* remove reference to HPC
  - See @fraserphysics comments for @sarostru
* rework the SQP discussion to be simplified
  - Our estimate of the Fisher information drops a term in the second
    derivative with the justification that its expected value is
    small.  We believe that it's an important point.
* Say if there is anything novel about the optimization algorithm
  - Added brief discussion of the difference from SQP
* Make the paper less domain specific
  - See @fraserphysics comments to @sarostru,
* Describe why different optimization methods were used i.e. brentq
  - The big optimization searches for the MAP estimate.  Our code for
    that goal is ad hoc and we plan to improve it in future work.  We
    use cvxopt inside our main loop because it provides linearly
    constrained optimization of a quadratic objective.  We use brentq
    to solve for the implicitly defined CJ conditions and don't
    consider that optimization.

sarostru commented:
The explanatory approach of your revised abstract and introduction seems very good to me, I think they succeed in clarifying your intent and making the work that follows accessible to a wider audience (myself included).  By clarifying the intent, I think that resolves my concerns about things being too domain specific, as they were largely since previously I thought you were trying to introduce a general purpose method.  I think this will still be a challenging read for non-specialists, but I feel like there is enough there that an interested party can make sense of the paper and appreciate the results.

I and several other reviewers commented on there not being a lot of discussion on python, in referring to the [review criteria] (https://github.com/scipy-conference/scipy_proceedings/blob/master/review_criteria.md) the relevant description is:
 > The work should describe the development or use of python software for approaching a problem in a domain within the scope of the conference.

I would still like to see this "use of python software" clause addressed more directly.  I think a small section that expands on the paragraph beginning with, "We are also using the project to learn and demonstrate..." would more firmly establish this objective.  You could address questions like: 
 - Why did you choose python for this?
  - Are there libraries or aspects of python that enable something that you couldn't otherwise do?
  - What about it enables your particular double-barreled HPC development process?
 - What about using python makes it particularly amenable to be used for, "a wide range of experiments and simulations"?

On Figure 8, since the values are so different at a glance it looks like there is only one eigenvalue, I think the other two are the boxes in the top left and bottom right, but I'm not sure.  I think the top figure should still make it clear there are 3 eigenvalues, you may want to use a table instead of a plot.   

With the additional python description, I would be comfortable merging this in.

fraserphysics commented:
Scott,

I just pushed changes in response to your email.  I also corrected some
BS that I had in the previous version about the second derivative of the
log likelihood and our calculation of Fisher information (things are
better than I had thought).

Here is how we respond to you comments:

>>>>> "SR" == Scott Rostrup <notifications@github.com> writes:

    SR> The explanatory approach of your revised abstract and
    SR> introduction seems very good to me, I think they succeed in
    SR> clarifying your intent and making the work that follows
    SR> accessible to a wider audience (myself included). [...]


Thanks.

    SR> I and several other reviewers commented on there not being a lot
    SR> of discussion on python, in referring to the review criteria the
    SR> relevant description is:

    SR>     The work should describe the development or use of python
    SR> software for approaching a problem in a domain within the scope
    SR> of the conference.

    SR> I would still like to see this "use of python software" clause
    SR> addressed more directly. I think a small section that expands on
    SR> the paragraph beginning with, "We are also using the project to
    SR> learn and demonstrate..." would more firmly establish this
    SR> objective. [...]

We did exactly that.

    SR> On Figure 8, since the values are so different at a glance it
    SR> looks like there is only one eigenvalue, I think the other two
    SR> are the boxes in the top left and bottom right, but I'm not
    SR> sure. I think the top figure should still make it clear there
    SR> are 3 eigenvalues, you may want to use a table instead of a
    SR> plot.

We changed the plot bounds and examined all the eigenvalues.  There are
only 4 nonzero eigenvalues and the new plot displays them all clearly.
The corresponding "eigenfunctions" are only nonzero near v_CJ.  I
expected a single significant eigenvalue with all others being 0 to
numerical precision.  We will have to figure out what's going on later.
At least the local nature of the "eigenfunctions" seems right.


    SR> With the additional python description, I would be comfortable
    SR> merging this in.

Thanks, and thanks to all reviewers for the care you have taken.  We
liked doing it all under version control.

Andy
-- 
Andy Fraser                                            FraserPhysics
197 Taos St                                            andy@fraserphysics.com
Los Alamos, NM 87544                                   505 695-2788

sarostru commented:
Great, we can let the other reviewers weigh in with their final thoughts as well.  Below is a quick timeline change update that looks like it won't affect you.

Due to the complication of the proceedings deadline lining up with everyone's travel schedule to attend the conference we've decided to allow authors to spend an additional 2 days, that is until July 12th to complete any major changes to the submission.  This may impact your enjoyment of the conference, so we hope not too many of you need to take the time.  For reviewers, we ask that you read over this final document before the conference is over and post your decision here in the comments on or before July 17th.

Thanks and have a great Scipy!
Your proceedings co-chairs Scott Rostrup (@sarostru) and Sebastian Benthall (@sbenthall)

rcpaffenroth commented:
The author's have addressed all of my comments from my review.   I have nothing additional to add and the paper is ready for merging.

certik commented:
It looks good to me too, my comments have been addressed.

+1 to merge as it is.

sarostru commented:
@fraserphysics Thanks for all your updates, looking great, I added the ready label.  At this point a final proofread is all we ask, since we don't have dedicated copy-editors, once that is done @sbenthall and I will help with updating the proceedings build from draft to final.

fraserphysics commented:
Scott,

We have done a final proofread, and I just pushed "Final version" to
https://github.com/fraserphysics/scipy_proceedings/tree/2016

I believe that is linked to our pull request, but I am a github novice.

Since you wrote the comment below, we have discovered a bug that put
incorrect eigenvalues in Figure 8.  We had expected a single positive
eigenvalue but the bug reported 4 numerically significant eigenvalues.
Having corrected the bug, the figure illustrates that there is only one
nonzero eigenvalue to within machine precision.  We have removed the
text that pointed out a discrepancy between the numerical results and
our expectations because there is no discrepancy.

Andy

>>>>> "SR" == Scott Rostrup <notifications@github.com> writes:

    SR> @fraserphysics Thanks for all your updates, looking great, I
    SR> added the ready label. At this point a final proofread is all we
    SR> ask, since we don't have dedicated copy-editors, once that is
    SR> done @sbenthall and I will help with updating the proceedings
    SR> build from draft to final.

    SR> — You are receiving this because you were mentioned.  Reply to
    SR> this email directly, view it on GitHub, or mute the thread.



-- 
Andy Fraser                                            FraserPhysics
197 Taos St                                            andy@fraserphysics.com
Los Alamos, NM 87544                                   505 695-2788

sarostru commented:
Resolving the merge conflicts and merging on a PR, https://github.com/scipy-conference/scipy_proceedings/pull/230

sarostru commented:
Merged the previous one into the wrong branch, trying again, https://github.com/scipy-conference/scipy_proceedings/pull/236
